cargo-quickinstall/src/lib.rs

```diff
@@ -33,6 +34,69 @@ pub struct CrateDetails {
     pub target: String,
 }
 
+/// Return (archive_format, url)
+fn get_binstall_upstream_url(target: &str) -> (&'static str, String) {
+    let archive_format = if target.contains("linux") {
+        "tgz"
+    } else {
+        "zip"
+    };
+    let url = format!("https://github.com/cargo-bins/cargo-binstall/releases/latest/download/cargo-binstall-{target}.{archive_format}");
+
+    (archive_format, url)
+}
+
+/// Attempt to download and install cargo-binstall from upstream.
+pub fn download_and_install_binstall_from_upstream(target: &str) -> Result<(), InstallError> {
+    let (archive_format, url) = get_binstall_upstream_url(target);
+
+    if archive_format == "tgz" {
+        untar(curl(&url)?)?;
+
+        Ok(())
+    } else {
+        assert_eq!(archive_format, "zip");
+
+        let (zip_file, zip_file_temp_path) = NamedTempFile::new()?.into_parts();
+
+        curl_file(&url, zip_file)?;
+
+        unzip(&zip_file_temp_path)?;
+
+        Ok(())
+    }
+}
+
+pub fn do_dry_run_download_and_install_binstall_from_upstream(
+    target: &str,
+) -> Result<String, InstallError> {
+    let (archive_format, url) = get_binstall_upstream_url(target);
+
+    let cargo_bin_dir = get_cargo_bin_dir()?;
+
+    if archive_format == ".tgz" {
+        Ok(format_curl_and_untar_cmd(&url, &cargo_bin_dir))
+    } else {
+        Ok(format!(
+            "temp=\"$(mktemp)\"\n{curl_cmd} >\"$temp\"\nunzip \"$temp\" -d {extdir}",
+            curl_cmd = prepare_curl_bytes_cmd(&url).formattable(),
+            extdir = cargo_bin_dir.display(),
+        ))
+    }
+}
+
+pub fn unzip(zip_file: &Path) -> Result<(), InstallError> {
+    let bin_dir = get_cargo_bin_dir()?;
+
+    process::Command::new("unzip")
+        .arg(zip_file)
+        .arg("-d")
```

Comment:
Can we rely on windows having an `unzip` command that behaves how we want?

Reply:
(curl and tar are there since 2017 - https://learn.microsoft.com/en-us/virtualization/community/team-blog/2017/20171219-tar-and-curl-come-to-windows )

Interestingly, I was doing some searching and couldn't find anything about an unzip command, but apparently windows' bsdtar is able to extract zipfiles - https://superuser.com/a/1473255

Reply:
`unzip` is a command comes with the `zip`.

Reply:
@alsuren The windows CI [successfully installed binstall via `unzip`](https://github.com/cargo-bins/cargo-quickinstall/actions/runs/3920707330/jobs/6702492079), though it failed during bootstraping.

Reply:
The CI is fixed, so I'm going to merge this.